docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,721 @@
+package software.amazon.awssdk.services.s3.transfer;
+
+/**
+ * The S3 Transfer Manager is a library that allows users to easily and
+ * optimally upload and downloads to and from S3.
+ * <p>
+ * The list of features includes:
+ * <ul>
+ *   <li>Parallel uploads and downloads</li>
+ *   <li>Bandwidth limiting</li>
+ *   <li>Pause and resume of transfers</li>
+ * </ul>
+ * <p>
+ * <b>Usage Example:</b>
+ * <pre>
+ * {@code
+ * // Create using all default configuration values
+ * S3TransferManager tm = S3TranferManager.create();
+ *
+ * // Using custom configuration values to set max upload speed to avoid
+ * // saturating the network interface.
+ * S3TransferManager tm = S3TransferManager.builder()
+ *         .configuration(TransferManagerConfiguration.builder()
+ *             .maximumUploadBytesSecond(32 * 1024 * 1024) // 32 MiB
+ *             .build()
+ *         .build();
+ * }
+ * </pre>
+ */
+public interface S3TransferManager {
+    /**
+     * DownloadObject an object identified by the bucket and key from S3 to the
+     * given file.
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate transfer
+     * DownloadObject myFileDownload = tm.downloadObject(BUCKET, KEY, Paths.get("/tmp/myFile.txt");
+     * // Wait for transfer to complete
+     * myFileDownload().completionFuture().join();
+     * }
+     * </pre>
+     *
+     */
+    default DownloadObject downloadObject(String bucket, String key, Path file) {
+        return downloadObject(DownloadObjectRequest.builder()
+                        .getObjectRequest(GetObjectRequest.builder()
+                                .bucket(bucket)
+                                .key(key)
+                                .build())
+                        .build(),
+                file);
+    }
+
+    /**
+     * DownloadObject an object using an S3 presigned URL to the given file.
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate transfer
+     * DownloadObject myFileDownload = tm.downloadObject(myPresignedUrl, Paths.get("/tmp/myFile.txt");
+     * // Wait for transfer to complete
+     * myFileDownload()completionFuture().join();
+     * }
+     * </pre>
+     */
+    default DownloadObject downloadObject(URL presignedUrl, Path file) {
```

Comment:
I didn't want to use the same names as the S3 client operations, so threw out `getObject`. In v1 we have `download` and `downloadDirectory`, but I felt `downloadObject` was more symmetric. I think this may be something we discuss with the other SDKs so we get consistency with any transfermanager-like libraries.

Reply:
I use `getItem` and `putItem` in the mapper, which matches the low-level client. I can understand the benefit in being different here, though.

I'm not too hung up on the verb, as long as we're consistent. I'm not a fan of 'Object', though. In the Java world it literally could mean anything.